### PR TITLE
feat(backend): New Backend and authentication for Phonebook data.

### LINKED ISSUE #18: Add build for Phonebook.Source.PeopleSoft
Currently, the project is not built by our CI Pipeline. 

Please add PR and production Builds under `.azure\pipelines``.

### LINKED ISSUE #271: Basic Auth
**Describe the bug:**
I will need to login again after restarting the computer. The **phonebook** looks like the first start.

Related to people backend.

**System Info:**
Chrome for Mac Version 75.0.37770.142

*Tfs Bug 8734*

### PR DESCRIPTION
**Pull request by paule96**
The following are the changes, most of them are breaking changes:

- replace the json file with a real backend
   - the backend connects directly to a database
   - currently, it try to return the same stuff as the old JSON file
- Open Source the backend
- add authentification to the people data
   - currently, supported is WsFederation over Azure or ADFS (and tested)

There are some issues that currently are known:

- the search for numbers is broken
   - this is not related to this change but good to know (#591 )
- the last hierarchy in the org structure isn't visible in any view
   - organigram
   - users supervisor
   - Users organigram information
- after login the user is redirected to `/` and not to the path the redirection starts. (#586 )
- the picture backend asked for login
   - this will change soon because we plan to use the same auth like this backend.

But we must ship this because the old source is broken. (after more than 15 years!)
So let's

ship it gif

BREAKING CHANGE
closes #18
closes #271

### CODE REVIEW

**Comment by paule96** [comment]
This pr is blocked by a missing build infrastructure! I add this later. @DanielHabenicht  we must do some internal discussions about this...

File: Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/BuildingController.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    [ApiController]
+    public class BuildingController : ControllerBase
+    {
+        public ModelContext Context { get; }
+
+        public BuildingController(ModelContext context)
+        {
+            Context = context;
+        }
+        // GET: api/Building
```
Comment (DanielHabenicht) [inline Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/BuildingController.cs]: Can you add valid XML markdown for integration of swashbuckle? 
Maybe we can also add swashbuckle in this pr or later on.

**Comment by paule96** [comment]
Good news @DanielHabenicht : This PR will continue on Monday :)

**Comment by paule96** [comment]
think about query caching with:

https://github.com/VahidN/EFSecondLevelCache.Core
or this:

https://github.com/SteffenMangold/EntityFrameworkCore.Cacheable/

things that are also todo is to add maybe OData or something like this to allow querying data.

**Comment by DanielHabenicht** [comment]
I think just provide the bare minimum for the first PR. Querying Data is for a later version as we already discussed.

**Comment by paule96** [comment]
Okay, @DanielHabenicht and @mschwrdtnr, I have e-mailed you the current JSON result of the API, so you can check if you missing something. The result for a dataset of 2000 employees has a response time of 500ms. If this is to slow I must find a way to cache it.

**Comment by Zeroks77** [comment]
https://stackoverflow.com/questions/807878/how-to-make-javascript-execute-after-page-load

**Comment by DanielHabenicht** [comment]
> https://stackoverflow.com/questions/807878/how-to-make-javascript-execute-after-page-load/36096571#36096571

Whats that?

**Comment by paule96** [comment]
> > https://stackoverflow.com/questions/807878/how-to-make-javascript-execute-after-page-load/36096571#36096571
> 
> Whats that?

That is maybe the solution for the out problem. But I'm not shure. 

So the idea is to open a new tab or window for the login. If the phonebook is the parent window we can watch in the DOM of this site. And maybe we are able to create an event that is triggered if the login is sucess. But my current plan is an other plan. Maybe we redirect the user always to the login. That is a breaking change to the current phonebook, because currently you can use some features without login. After this you can't. So I think I start with the easy one. So we can start finishing this PR.

**Comment by paule96** [comment]
@T-Systems-MMS/phonebook-developers there new updates for this branch. It's now working with the frontend. But there are two bugs:

- the rooms in the rooms view are empty
- the owner of orgs in the orgstructurs highest level are empty

**Comment by paule96** [comment]
@DanielHabenicht the frontend is a little bit broken. So maybe you can fix the compile errors.

**Comment by DanielHabenicht** [comment]
431 File changes 😭

**Comment by paule96** [comment]
> 431 File changes 😭

What?

image

**Comment by DanielHabenicht** [comment]
Der Filter auf 2 commits war noch drin 👍

**Comment by DanielHabenicht** [comment]
Gibt's ne Möglichkeit dein Backend ohne Datenverbindung zu testen?

**Comment by paule96** [comment]
No currently not.

File: Phonebook.Frontend/src/app/modules/table/helpers.ts [hunk 2]
```diff
@@ -94,6 +94,9 @@ export class Helpers {
    * @param str Any String
    */
   public static removeAccents(str: string): string {
+    if(str ==null){
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/modules/table/helpers.ts]:
```suggestion
  public static removeAccents(str: string | null): string {
    if(str === null){
```
Bitte die Funktioneingabe dann auch anpassen

Comment (DanielHabenicht) [reply Phonebook.Frontend/src/app/modules/table/helpers.ts]: Obwohl das eigentlich vorher behoben werden sollte.

File: Phonebook.Frontend/src/app/shared/pipes/telephone.pipe.ts [hunk 3]
```diff
@@ -5,6 +5,6 @@ import { Pipe, PipeTransform } from '@angular/core';
 })
 export class TelephonePipe implements PipeTransform {
   public transform(telephoneNumber: string): string {
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/shared/pipes/telephone.pipe.ts]: Auch hier bitte die Funktion und Ausgangswerte im Model mit anpassen! 
```suggestion
  public transform(telephoneNumber: null | string): string {
```

File: Phonebook.Frontend/src/app/modules/table/helpers.ts (hunk 2 above)
Comment (DanielHabenicht) [reply Phonebook.Frontend/src/app/modules/table/helpers.ts]: oder im Modell

File: Phonebook.Frontend/src/app/shared/interceptors/HttpRedirectToLogin.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import { map, catchError } from 'rxjs/operators';
+
+@Injectable({ providedIn: 'root' })
+export class HttpRedirectToLogin implements HttpInterceptor {
+  constructor(private http: HttpClient) { }
+
+  public intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    return next.handle(req).pipe(
+      map(d => {
+        if (d instanceof HttpResponse && d.status == 401 && d.headers.has("Location")) {
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/shared/interceptors/HttpRedirectToLogin.ts]: Evtl. sollten wir das noch etwas einschränkender machen und nur für die Domain auf der das Backend läuft ausführen

**Comment by DanielHabenicht** [comment]
@Compufreak345  willst du mal über den .net Code drübergucken?

File: Phonebook.Frontend/src/app/shared/interceptors/HttpRedirectToLogin.ts (hunk 4 above)
Comment (paule96) [reply Phonebook.Frontend/src/app/shared/interceptors/HttpRedirectToLogin.ts]: das geht nicht, da durch unseren Proxy alles die selbe Domain hat :D

Comment (DanielHabenicht) [reply Phonebook.Frontend/src/app/shared/interceptors/HttpRedirectToLogin.ts]: Ich guck dann nochmal rein. Das ist mir im Moment klar, aber vielleicht haben wir in zukunft noch andere domains z.B. den bilderservice.

**Comment by DanielHabenicht** [comment]
Thinigs still to do

- [ ] ask use before navigating
- [ ] navigate to page where the user left of.

**Comment by paule96** [comment]
no it is testable localy. I will next try to fix the build and the preview

**Comment by paule96** [comment]
now you can simply run `npm run start -- --open` like always to test the new backend.

There are currently some things left:

- [ ] Add a cleaner generator for fake data. Maybe [bogus](https://github.com/bchavez/Bogus)
- [ ] remove our WS federation information
- [ ] make the mocked version be able to work without out auth -> to make development easier
- [ ] cleanup the null warnings -> 111 currently 👀
- [ ] write documentation about all of the
- [ ] please test all buttons

**Comment by DanielHabenicht** [comment]
Please merge maser into this branch to deactivate restyled.

**Comment by paule96** [comment]
done

**Comment by DanielHabenicht** [comment]
The end to end test is literally failing because of the first april :D

File: .azure/pipelines/pr/Phonebook.Source.PeopleSoft.pr.yml [hunk 5]
```diff
@@ -1 +1,19 @@
 # TODO: #18
```
Comment (DanielHabenicht) [inline .azure/pipelines/pr/Phonebook.Source.PeopleSoft.pr.yml:1]:
```suggestion
```

File: Phonebook.Frontend/src/app/modules/rooms/components/room/room.component.ts [hunk 6]
```diff
@@ -55,7 +55,7 @@ export class RoomDetailComponent implements OnInit {
             this.room = this.node.data as Room;
           }
           this.personService
-            .getPersonsByRoom(RoomHelpers.getParamsAsArray(params, ['cityId', 'buildingId', 'floorId', 'roomId']))
+            .getPersonsByRoom(RoomHelpers.getParamsAsArray(params, ['roomId'])[0])
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/modules/rooms/components/room/room.component.ts:58]: Das funktioniert nicht, es gibt Räume die die gleiche RoomId haben, oder mittlerweile nicht mehr?

File: Phonebook.Frontend/src/app/modules/table/helpers.ts [hunk 7]
```diff
@@ -93,7 +93,10 @@ export class Helpers {
    * Removes Accents from the string.
    * @param str Any String
    */
-  public static removeAccents(str: string): string {
+  public static removeAccents(str: string | null): string {
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/modules/table/helpers.ts:96]: das verursacht evtl. unbeabsichtigtes fehlverhalten, da `removeAccents` in sortierfunktionen eingesetzt wird. Ich bau es mal so um, dass es trotzdem funktioniert.

Ich hoffe du hast es die nullbaren Dinge in den Models angepasst?

File: Phonebook.Frontend/src/app/services/api/person.service.ts [hunk 8]
```diff
@@ -31,7 +31,7 @@ export class PersonService {
           item.Contacts.Fax,
           item.Contacts.Email,
           item.Contacts.Phone,
-          new Messenger(item.Contacts.Messenger.Text, item.Contacts.Messenger.State)
+          new Messenger("", null)
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/services/api/person.service.ts:34]: Da steht nie etwas drin? 
Es wird auch im Phonebook nicht genutzt... Soll ich es entfernen?

File: Phonebook.Frontend/src/app/services/api/person.service.ts [hunk 9]
```diff
@@ -102,15 +103,12 @@ export class PersonService {
     );
   }
 
-  public getPersonsByRoom(positionArray: string[]): Observable<Person[]> {
+  public getPersonsByRoom(roomId: string): Observable<Person[]> {
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/services/api/person.service.ts:106]: gleiches problem roomId war vorher nicht unique

File: Phonebook.Frontend/src/app/shared/components/user/user-detail/user-detail.component.html [hunk 10]
```diff
@@ -3,34 +3,25 @@
     <mat-card-header>
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/shared/components/user/user-detail/user-detail.component.html:3]: Was hast du in dieser Datei geändert? Ich finde es nicht.

File: Phonebook.Frontend/src/app/shared/models/classes/Messenger.ts [hunk 11]
```diff
@@ -3,8 +3,8 @@ export class Messenger {
   public Text?: any;
   public State: number;
 
-  constructor(text: any = null, state: number) {
+  constructor(text: any = null, state: number | null) {
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/shared/models/classes/Messenger.ts:6]: remove if unused

File: Phonebook.Frontend/src/app/shared/pipes/telephone.pipe.ts [hunk 12]
```diff
@@ -4,7 +4,7 @@ import { Pipe, PipeTransform } from '@angular/core';
   name: 'telephone'
 })
 export class TelephonePipe implements PipeTransform {
-  public transform(telephoneNumber: string): string {
-    return telephoneNumber.replace(/[^+0-9]/g, '');
+  public transform(telephoneNumber: null | string): string {
```
Comment (DanielHabenicht) [inline Phonebook.Frontend/src/app/shared/pipes/telephone.pipe.ts:7]:
```suggestion
TODO: #579
  public transform(telephoneNumber: null | string): string {
```

File: Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft.Models/Old/City.cs [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+                }
+                if (room.BuildingPart != null && room.BuildingPart.Building != null && room.BuildingPart.Building.Name != null)
+                {
+                    return room.BuildingPart.Building.Name;
+                }
+                if (room.Floor != null && room.Floor.Building != null && room.Floor.Building.Name != null)
+                {
+                    return room.Floor.Building.Name;
+                }
+                // TODO: add a warn!
```
Comment (DanielHabenicht) [inline Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft.Models/Old/City.cs:53]: todo?

File: Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/appsettings.Development.json [hunk 14]
```diff
@@ -5,5 +5,7 @@
       "System": "Information",
       "Microsoft": "Information"
     }
-  }
-}
\ No newline at end of file
+  },
+  "useSeeding": true,
+  "useAnonymos":  true
```
Comment (DanielHabenicht) [inline Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/appsettings.Development.json:10]: typo
```suggestion
  "useAnonymous":  true
```

File: global.json [hunk 15]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (DanielHabenicht) [inline global.json:1]: Is this the reight spot for this file?

File: Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/BuildingController.cs (hunk 1 above)
Comment (DanielHabenicht) [reply Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/BuildingController.cs:22]: --> Issue: #585

File: Phonebook.Frontend/src/app/modules/rooms/components/room/room.component.ts (hunk 6 above)
Comment (paule96) [reply Phonebook.Frontend/src/app/modules/rooms/components/room/room.component.ts:58]: test before close. The room Id is now the DatabaseId, thats mean it is unique.

File: Phonebook.Frontend/src/app/modules/table/helpers.ts (hunk 7 above)
Comment (paule96) [reply Phonebook.Frontend/src/app/modules/table/helpers.ts:96]: set all model properties to nullable.

File: Phonebook.Frontend/src/app/services/api/person.service.ts (hunk 8 above)
Comment (paule96) [reply Phonebook.Frontend/src/app/services/api/person.service.ts:34]: ja!!!!!! bitte!

File: Phonebook.Frontend/src/app/services/api/person.service.ts (hunk 9 above)
Comment (paule96) [reply Phonebook.Frontend/src/app/services/api/person.service.ts:106]: is now unique

File: Phonebook.Frontend/src/app/shared/components/user/user-detail/user-detail.component.html (hunk 10 above)
Comment (paule96) [reply Phonebook.Frontend/src/app/shared/components/user/user-detail/user-detail.component.html:3]: nothing. hit save

File: global.json (hunk 15 above)
Comment (paule96) [reply global.json:1]: yes

File: Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/PeopleController.cs [hunk 16]
```diff
@@ -66,21 +66,21 @@ private IEnumerable<Person> InlcudeDependencies(IQueryable<Person> query)
                     new Person()
                     {
                         Id = p.Id,
-                        EMail = p.EMail,
-                        FAX = p.FAX,
-                        FirstName = p.FirstName,
+                        EMail = p.EMail == null? string.Empty: p.EMail,
```
Comment (DanielHabenicht) [inline Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/PeopleController.cs:69]: Nein! Lass mich das im Frontend machen.

Comment (paule96) [reply Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/PeopleController.cs:69]: dann ist das Frontend jetzt aber broken. Also entweder du baust das jetzt noch ein oder wir mergen das erstmal so.

**Comment by paule96** [comment]
@T-Systems-MMS/phonebook-developers please look at this. I will merge this at 7pm. If you find any issues that is your last chance ;)

File: Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/PeopleController.cs (hunk 16 above)
Comment (DanielHabenicht) [reply Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/PeopleController.cs:69]: nagut, ich erstelle ne issue

Comment (DanielHabenicht) [reply Phonebook.Source.PeopleSoft/src/Phonebook.Source.PeopleSoft/Controllers/PeopleController.cs:69]: #593

**Comment by DanielHabenicht** [comment]
ship it gif

**Comment by paule96** [comment]
Can we hack img tags in the Release Notes message? :D I wan't clippy in the release notes :D

**Comment by DanielHabenicht** [comment]
Please add before the "closes" the literal "BREAKING CHANGE" otherwise it won't get picked up by semantic release.

**Pull request merged**